Fix BackButton handling input before all other UI elements

nyquillerium opened the pull request:
Fixes https://github.com/ppy/osu/issues/5177

Attempt 2, now with 100% more automated testing.

This approach just rebuilds the `KeyBindingInputQueue` with the back button at the end.

nyquillerium committed:
Remove and re-add backbutton instead and add tests

iiSaLMaN commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,152 @@
+                        return true;
+
+                    default:
+                        current.Exit();
```
```suggestion
                        current.Exit();
                    case null:
                    case Intro _:
                    case Disclaimer _:
```

shortens the switches

nyquillerium replied:
I don't think this works unless I'm mistaken, but I can probably simplify this switch another way.

iiSaLMaN commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,152 @@
+        [Test]
+        public void TestExitingSongSelectWithEscape()
+        {
+            TestSongSelect songSelect = null;
```
redundant assign, same for the other tests

nyquillerium replied:
required because test steps may be manually ran out of order, which results in a warning that this value may be used without being assigned.

iiSaLMaN commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,152 @@
+            TestSongSelect songSelect = null;
+
+            AddStep("Push songselect", () => osuGame.ScreenStack.Push(songSelect = new TestSongSelect()));
+            AddUntilStep("Wait for song select", () => songSelect.IsCurrentScreen());
```
move to a function, probably `pushScreenStep(new TestSongSelect())`

iiSaLMaN commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,152 @@
+        {
+            TestSongSelect songSelect = null;
+
+            AddStep("Push songselect", () => osuGame.ScreenStack.Push(songSelect = new TestSongSelect()));
```
inconsistent name, add a space between `song` `select` *might be unnecessary*

nyquillerium committed:
Fix step names

smoogipoo commented:
On discussion with @peppy, it might be cleaner to do this via the action receptor model.

```yaml
Children:
    backButtonReceptor
    screenStack
    backButton(backButtonReceptor)

class BackButton:
    ctor(BackButtonReceptor receptor = null):
        receptor.OnPressed += ...
        receptor.OnReleased += ...
```

nyquillerium committed:
Use a receptor model instead

peppy commented on osu.Game/OsuGame.cs:
```diff
@@ -81,10 +81,13 @@ public class OsuGame : OsuGameBase, IKeyBindingHandler<GlobalAction>
+        protected OsuScreenStack ScreenStack;
+        protected BackButton BackButton;
+
+        private BackButton.BackButtonReceptor backButtonReceptor;
```
is there a reason this isn't a local?

peppy commented on osu.Game/Graphics/UserInterface/BackButton.cs:
```diff
@@ -61,5 +50,24 @@ protected override void PopOut()
             button.FadeOut(400, Easing.OutQuint);
         }
+
+        public class BackButtonReceptor : Drawable, IKeyBindingHandler<GlobalAction>
```
Since this is nested, just call it `Receptor`

peppy commented on osu.Game/Graphics/UserInterface/BackButton.cs:
```diff
@@ -10,14 +10,16 @@
-        public BackButton()
+        public BackButton(BackButtonReceptor receptor)
         {
+            receptor.OnBackPressed += () => Action.Invoke();
```
```suggestion
            receptor.OnBackPressed = () => Action.Invoke();
```

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,152 @@
+        public void SetUpSteps()
+        {
+            AddUntilStep("wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("wait for main menu", () =>
```
this isn't "waiting", this is "exiting"

nyquillerium committed:
Move screen pushes into function, rename receptor

nyquillerium committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

nyquillerium committed:
fix merge

nyquillerium committed:
update names

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneExitingScreens.cs:
```diff
@@ -0,0 +1,151 @@
+        public void SetUpSteps()
+        {
+            AddUntilStep("wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("exit to main menu", () =>
```
what's the purpose of this step? how would we be anywhere else?

nyquillerium replied:
We could potentially still be in the intro or disclaimer, and this makes sure you can reset a test in visual test browser even if you stop in the middle of one.

peppy replied:
that's not how it should be done. "resetting" should be making a new game. you're just asking for this to fail in random ways.

nyquillerium replied:
I guess we should create a game instance that skips directly to the main menu on run to avoid this step entirely.

peppy replied:
the step is fine. it shouldn't be calling exit

nyquillerium committed:
rename test scene and create a new game each test

nyquillerium committed:
fix step name and rename test

smoogipoo commented:
Exiting from overlays is not tested.

nyquillerium committed:
fix tests

nyquillerium committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

nyquillerium committed:
get tests running again

nyquillerium committed:
Add new options test

peppy commented on osu.Game/Screens/Multi/Multiplayer.cs:
```diff
@@ -166,7 +168,7 @@ public void Start(Func<Player> player)
         public void APIStateChanged(IAPIProvider api, APIState state)
         {
-            if (state != APIState.Online)
+            if (RequireOnline && state != APIState.Online)
```
can't this instead be done via the dummy api access?

nyquillerium committed:
Use dummyAPI and move load check into MusicController

nyquillerium committed:
Remove no longer needed property

nyquillerium committed:
Remove usage of no longer needed property

peppy commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -316,6 +316,10 @@ private void prev()
 
         private void next(bool instant = false)
         {
+            // beatmapSets doesn't get populated until loading has completed.
```
why is this here? if this is causing a crash or issue it should 100% be its own PR with own tests.

nyquillerium replied:
this is causing a crash only on headless runs when going through the load process (since loading the `MusicController` can take a while). Can split this out.

peppy replied:
very highly doubt it's headless only. nothing stops a desktop run from also taking equally as long.

nyquillerium replied:
Turns out this was unnecessary and was an issue in my test

nyquillerium committed:
Explicity dispose osuGame instances

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs:
```diff
@@ -0,0 +1,193 @@
+        [Test]
+        public void TestOpenOptionsAndExitWithEscape()
+        {
+            AddUntilStep("Wait for options to load", () => osuGame.Settings.IsLoaded);
```
this shouldn't be required.

nyquillerium replied:
This is required because this actually takes quite a long time in headless, and trying to toggle the options overlay off while its not loaded with escape will not work since the `FocusedTextBox` has not taken focus yet.

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs:
```diff
@@ -0,0 +1,193 @@
+            AddAssert("Options overlay was closed", () => osuGame.Settings.State.Value == Visibility.Hidden);
+        }
+
+        private void pushAndConfirm(Func<Screen> newScreen, string screenName)
```
you don't need the `string` argument

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs:
```diff
@@ -0,0 +1,193 @@
+            });
+            AddUntilStep("Wait for load", () => osuGame.IsLoaded);
+            AddUntilStep("Wait for intro", () => osuGame.ScreenStack.CurrentScreen is IntroScreen);
+            AddUntilStep("Wait for main menu", () => osuGame.ScreenStack.CurrentScreen is MainMenu menu && menu.IsLoaded);
```
this line is in 3 places

peppy commented on osu.Game.Tests/Visual/Menus/TestSceneScreenNavigation.cs:
```diff
@@ -0,0 +1,193 @@
+        private GameHost gameHost;
+        private TestOsuGame osuGame;
+
+        private Vector2 backButtonPosition => osuGame.ToScreenSpace(new Vector2(25, osuGame.LayoutRectangle.Bottom - 25));
```
you can make 25 a constant `click_padding` or something.

nyquillerium committed:
remove string param, move menu check to method, add const padding

nyquillerium committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

peppy committed:
Merge branch 'master' into back-button-part-2

nyquillerium committed:
add back missing loaded check

peppy committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

peppy committed:
Merge branch 'master' into back-button-part-2

peppy committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

peppy committed:
Use isolated storage/api

peppy committed:
Change intro displayed for tests

As the triangles intro relies on the audio track's clock advancing, we can't use it just yet (CI server has no audio device). This is a temporary workaround for that shortcoming.

smoogipoo commented on osu.Game/OsuGame.cs:
```diff
@@ -96,14 +100,14 @@ public class OsuGame : OsuGameBase, IKeyBindingHandler<GlobalAction>
 
         private readonly List<OverlayContainer> visibleBlockingOverlays = new List<OverlayContainer>();
 
+        protected virtual Loader CreateLoader() => new Loader();
```
Method should not be above ctor

smoogipoo commented on osu.Game/Graphics/UserInterface/BackButton.cs:
```diff
@@ -61,5 +50,24 @@ protected override void PopOut()
+                switch (action)
+                {
+                    case GlobalAction.Back:
+                        OnBackPressed.Invoke();
```
```suggestion
                        OnBackPressed?.Invoke();
```

smoogipoo commented on osu.Game/Graphics/UserInterface/BackButton.cs:
```diff
@@ -10,14 +10,16 @@
-        public BackButton()
+        public BackButton(Receptor receptor)
         {
+            receptor.OnBackPressed = () => Action.Invoke();
```
```suggestion
            receptor.OnBackPressed = () => Action?.Invoke();
```

peppy committed:
Merge remote-tracking branch 'upstream/master' into back-button-part-2

peppy committed:
Move protected method

peppy committed:
Add invocation null checks for safety

peppy committed:
Merge branch 'master' into back-button-part-2
